### PR TITLE
Jenkins & Github actions

### PR DESCRIPTION
**Pull request by cosunae**
This PR introduces new github actions and jenkinsfile in order to regularly test the main branch and commits in a PR. 
The intended overall strategy is the following: 
# Github Actions
* github actions create conda environments with requirement.in and dev-requirement.in in order to test the environments. Additionally it will run pre-commit on the blueprint.
* actions will appear in the actions tab of the repo https://github.com/MeteoSwiss-APN/mch-python-blueprint/actions
* any commit in the main triggers the github actions. 
* any commit in a PR triggers the github actions so that PR can not be merged into main if actions do not pass
# Jenkins
* jenkins will execute the cookiecutter to create a default project, install the requirements and run the tests. 
* jenkins job can be triggered from any PR by writing the comment 'launch jenkins' in the PR. Therefore jenkins must be triggered explicitly. This is on purpose since tests can be expensive and it is not desirable to run on every commit. 
The jenkins plan for testing PR is https://jenkins-mch.cscs.ch/job/blueprint_PR/
* jenkins has a nightly build plan that runs (same as for the PR plan) regularly. The aim of the regular tests is to capture changes in unpinned environments that might break the environment even if there are no code modifications. 
The regular plan for testing is https://jenkins-mch.cscs.ch/job/blueprint/

# Functionalities and modifications:
* For a PR, both the github actions and jenkins tests are reporting the results in the PR, right above the merge buttom (see this same PR for an example).
* remove the sample_code option from the cookiecutter templates in order to reduce the dimensionality. (we could not test all options). We kept the calculator sample as the only template.
* jenkinsfile for testing the blueprint (as described above): https://github.com/MeteoSwiss-APN/mch-python-blueprint/pull/29/files#diff-6c263717a24af2836fc1795f59bdea6b8616d61bd499edb331d6c9d253478e34
The same is used for the regular tests and the PR testing. 
* jenkinsfile inside the cookiecutter project template: 
https://github.com/MeteoSwiss-APN/mch-python-blueprint/pull/29/files#diff-2cf7a01fd054aeab9eba10da71988b50e88bf395098e3e2abd9630a77a114fc4
It mimics the one above in the blueprint (without the cookiecutter project generation). Currently this can not be tested since there is no instance of a cookiecutter project. 
* a jenkins job generation code: https://github.com/MeteoSwiss-APN/mch-python-blueprint/pull/29/files#diff-d10ff04c672ba85d4d4cfd6457212806c6f35b32c617f1f189546d51c0e5ade3
This can be used in order to automatically create and configure the job PR plan in jenkins. We should document how to use it. 
* update the pre-commit environment (some version like rstcheck and sphinx were old and do not build anymore): 
https://github.com/MeteoSwiss-APN/mch-python-blueprint/pull/29/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R3
* rstcheck disabled for the blueprint: 
currently there are many code blocks without a language, and when we add a language, things like 
.. code:: bash
  cd <workdir>
break the checker. Although the language is optional the checker does not allow not setting it. 
https://rstcheck-core.readthedocs.io/en/latest/faq/#code-blocks-without-language-sphinx

Missing: check the pinned environment.

### CODE REVIEW

**Comment by cosunae** [comment]
launch jenkins

**Review by ruestefa — COMMENTED** [review]
Some minor suggestions in documentation files, otherwise looks good. The Jenkins file I cannot judge as I'm not familiar with it.

**Comment by ruestefa** [comment]
I like that the github actions use the unpinned dependencies, which ensures that there are not surprises when updating the pinned environments to the latest versions. I'm assuming Jenkins will then test the (operational) pinned environments? Will github actions use pip (i.e., pypi) or conda, though? If the former, there may still be surprises when updating the conda environments as some packages are usually behind pypi in version.

File: readme/installation_and_dependencies.rst
Comment (cosunae) [reply readme/installation_and_dependencies.rst]: I replaced all the bashes here by shell. That was a good one, actually I think I could recover the rst checks that were broken because of this

File: {{cookiecutter.project_slug}}/requirements/requirements.in
Comment (ruestefa) [reply {{cookiecutter.project_slug}}/requirements/requirements.in:2]: There are good arguments for either way. Keeping the test environment close as close to the runtime environment as possible definitely seems sensible. The cleanest might be to add a `test-requirements.in` (and consequently a `test-environment.yml`), though whether adding additional files would be worth it over pragmatically adding `pytest` to `requirements.in` is questionable, at least as long it's only `pytest`...

For the actual runtime environment (`environment.yml`) it probably wouldn't make a difference in most cases, anyway, as chances are that some project in the dependency tree depends on `pytest` (as is, e.g., the case for [`pyflexplot`](https://github.com/MeteoSwiss-APN/pyflexplot/blob/master/environment.yml)).

I guess I would keep your change (i.e., `pytest` in `requirements.in`) and document the reasoning behind it, with a remark that in case of more complex testing setups that require multiple packages, adding `test-requirements.in` and `test-environment.yml` may be sensible.

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (ruestefa) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: @regDaniel Indeed, the `*requirements.txt` are essentially replaced by `*environment.yml` files with the switch to conda and can thus be removed. (Doesn't necessarily have to be done in this PR, though.)

File: {{cookiecutter.project_slug}}/requirements/requirements.in
Comment (regDaniel) [reply {{cookiecutter.project_slug}}/requirements/requirements.in:2]: I would second your opinions here and go for the pragmatic solution, leave the test requirements (i.e. at the moment only pytest) in `requirements.in`. The created overhead in the runtime environment is likely marginal but it keeps the worklflow easier when we switch from pinned to unpinned environments.

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (pirmink) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: Just updated a PR for pytrajplot today where I very much reduced the Makefile to only the essential tasks. Here my experience concerning requirement specification, which might have interesting implications for the blueprint as well, especially regarding the name of the requirements file:

- `conda create` only accepts a simple list of packages in a text file (exactly one package name and, if necessary, the version specification per line, no dash, no lines with colon) and fails with yaml files
- `conda env create` on the contrary accepts both yaml and simple lists but requires the input file to have either a .yaml, .yml or .txt extension. All others file extensions, namely requirements**.in**, are rejected!

My conclusion was to have a requirements.txt (not requirements.in) file with unpinned requirements and an environment.yml file for pinned requirements, both for use with `conda env create`. 

BTW For pytrajplot, the appropriate file is chosen depending on the `PINNED ` variable that can be set to 0 for unpinned and to 1 (the default) for pinned requirements in the call to make, as in `make install PINNED=0`. However as you are moving away from make, this is probably out of question for the blueprint.

Comment (ruestefa) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: As far as I understand it, the purpose of `conda env create` is to exactly recreate existing environments, hence it's acceptence of `*environment.yml` and `*requirements.txt` files, which are the standard names for files with pinned dependencies. By contrast, `conda create` is the general-purpose way of creating an environment and thus accepts any files with dependencies, whether they can be expected be pinned or not. (If I remember correctly, I once read a discussion in a conda issue that concluded that `conda env create` is pretty much only retained for backward compatibility, though it's been a while don't have the link at hand...)

> My conclusion was to have a requirements.txt (not requirements.in) file with unpinned requirements

Since dependency specification in Python is very much a jungle, the `*requirements.in` naming is not a ubiquitous standard (but at least the way a fairly widely used tool (`pip-tools` I think?) does it this way). But what's for sure standard is using `requirements.txt` for pinned dependencies, so I wouldn't recommend using `.txt` for files with unpinned dependencies, as this goes actively against about the only standard there is.

File: .github/workflows/precommit.yml
Comment (ruestefa) [inline .github/workflows/precommit.yml:39]: Here, the runtime requirements in `requirements.in` should be installed before those in `dev-requirements.in`, shouldn't they? Otherwise, you only have the dev tools but are missing the actual runtime requirements...

**Comment by cosunae** [comment]
launch jenkins

**Comment by cosunae** [comment]
> I like that the github actions use the unpinned dependencies, which ensures that there are not surprises when updating the pinned environments to the latest versions. I'm assuming Jenkins will then test the (operational) pinned environments? Will github actions use pip (i.e., pypi) or conda, though? If the former, there may still be surprises when updating the conda environments as some packages are usually behind pypi in version.

When you reviewed it, the test of the pinned environments was missing. I added jenkins stages for all combinations: req, dev-req, pinned-req, and pinned-dev-req. Please have a look and comment on that. 
Concerning the installation, both github actions and jenkins are using the same using conda.

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (cosunae) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: ok, so the conclusion is we still need requirements.in (for unpinned) and requirements.txt (for pinned)? 
Not sure if there is consensus here, but perhaps we can shortly discuss it otherwise sometime next week.
For the time being and for this PR, I didnt change the convention. I kept it as it is, but updated the pinned requirements (that were not working), and added all the tests in jenkins

**Comment by cosunae** [comment]
launch jenkins

File: .github/workflows/precommit.yml
Comment (cosunae) [reply .github/workflows/precommit.yml:40]: good catch, that was a missunderstanding of the conventions from my side. 
I understood that dev-requirements should contain all dependencies of requirements.
I changed it in both github actions and jenkins

Comment (ruestefa) [reply .github/workflows/precommit.yml:40]: > I understood that dev-requirements should contain all dependencies of requirements.
No, that only applies to `dev-environment.yml`, which is a superset of `environment.yml`

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (regDaniel) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: I do not think that we still need `requirements.txt` for the pinned installation or at least I would not see a reason why. If it's just about the env name for `conda create`, we'll find another way as the env name is in principle set by cookiecutter (or manually by the user) upon its first creation.

Comment (regDaniel) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: follow-up: I see that the jenkins plan uses requirements.txt files for pinned installations. Of course it's also possible to drop the .yml files, i the generation of a requirements.txt is more straightforward or has other advantages, I guess this is up to @ruestefa to decide once he writes the env export script. But we should decide between the two and keep the installation process the same for jenkins and for the actual installation with the installation scripts.

File: {{cookiecutter.project_slug}}/.github/workflows/precommit.yml
Comment (regDaniel) [inline {{cookiecutter.project_slug}}/.github/workflows/precommit.yml:6]: Change to `main`.

File: {{cookiecutter.project_slug}}/.github/workflows/precommit.yml
Comment (regDaniel) [inline {{cookiecutter.project_slug}}/.github/workflows/precommit.yml:9]: Change to `main`.

**Review by regDaniel — CHANGES_REQUESTED** [review]
Minor issues only. Please check the `precommit.yml` files carefully to have the correct branch names before merging.

I will adjust the jenkins plan to use the installation scripts once this PR is merged, ok @cosunae ?

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (cosunae) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: ok, let me see if I can summarize, @ruestefa  argues requirements.yml or requirements.txt for pinned dependencies and requirements.in for unpinned dependencies (since a more standard in pip).  @regDaniel argues for having requirements.yml for pinned dependencies. So I can move the requirements.txt to requirements.yml. 
The only reason @regDaniel the jenkins files are using `requirements.txt` is because   `requirements.yml` do not exist yet (as far as I know, isnt?) 

One more question, since we are using conda, and not pip, why not using requirements.yml then for both, pinned and unpinned '

File: .github/workflows/precommit.yml
Comment (cosunae) [reply .github/workflows/precommit.yml]: thanks, yes, I already changed it

File: {{cookiecutter.project_slug}}/.github/workflows/precommit.yml
Comment (cosunae) [reply {{cookiecutter.project_slug}}/.github/workflows/precommit.yml]: yes, thanks. done. Also it could be that the default branch is named something else by the user. In that case either we add it to the templates of the cookicutter or the user has to change it by hand

File: {{cookiecutter.project_slug}}/.github/workflows/precommit.yml
Comment (cosunae) [reply {{cookiecutter.project_slug}}/.github/workflows/precommit.yml]: done

File: {{cookiecutter.project_slug}}/requirements/requirements.txt
Comment (ruestefa) [reply {{cookiecutter.project_slug}}/requirements/requirements.txt:1]: > since we are using conda, and not pip, why not using requirements.yml then for both, pinned and unpinned

The `*.yml` files are created with `conda env export`, which exports the whole environment (full dependency tree with pinned version numbers). I am not aware of a way to specify unpinned top-level dependencies in `*.yml`. Presumably, one could manually write a `*.yml` file following the format that conda exports and conda may accept that during environment creation, but this is really needlessly complicated as a plain text file is more human-friendly than yaml. There's also never any need to manually edit an `*environment.yml` file, so there's no benefit in using the same format for unpinned requirements and pinned environments.

**Review by regDaniel — APPROVED** [review]

**Comment by cosunae** [comment]
For reference, I added a last commit to update the .pre-commit config in the template (it was updated only in the parent blueprint project, I forgot it in the templates) and a couple of small fixes in the doc. 
I generated a default flying_circus project, pushed it in a new github repo, and tested there the github actions and jenkins file. All green, but these tests were done by hand.

**Pull request merged**
